### DISCUSSION TITLE
plot the spectral function and self energy with svo.h5

### DISCUSSION THREAD

**Discussion by KeyuAN95** [TRIQS/solid_dmft#92] [Q&A]
Dear Alex and all,

Question1:
I want to plot the spectral function with this tutorial [https://triqs.github.io/solid_dmft/tutorials/correlated_bandstructure/plot_correlated_bands.html](url).
I can run the Python script successfully using the svo_example.h5 file.[https://github.com/TRIQS/solid_dmft/tree/3.3.x/doc/tutorials/correlated_bandstructure](url).

However, there is an error if I run the solid_dmft with these input documents to get svo.h5. [https://github.com/TRIQS/solid_dmft/tree/3.3.x/doc/tutorials/SVO_os_qe](url)
[input.zip](https://github.com/user-attachments/files/18076685/input.zip)
![Snipaste_2024-12-10_17-31-27](https://github.com/user-attachments/assets/05ecb936-0311-44a6-a87d-e4fcd43f66bb)

The dmft_config.toml is the following :
#############
[general]
seedname = "svo"
jobname = "out"
enforce_off_diag = true
block_threshold = 0.001
prec_mu = 0.001
h_int_type = "kanamori"
U = 2.0
J = 0.65
mu_initial_guess = 12.297745
beta = 40
n_iter_dmft = 8
dc_type = 1
dc = true
dc_dmft = false
calc_energies = false
sigma_mix = 1.0
h5_save_freq = 2
[solver]
type = "cthyb"
n_l = 35
length_cycle = 120
n_warmup_cycles = 8000
n_cycles_tot = 10e+6
imag_threshold = 1e-5
measure_G_l = true
perform_tail_fit = false
############
So, how can I  get an accurate svo.h5 like svo_example.h5?
Question2:
I tried to plot the self-energy of the above svo.h5 with the 01s-solid_dmft.py in this tutorial. [https://github.com/TRIQS/tutorials/blob/3.3.x/AbinitioDMFT/solutions/01s-solid_dmft.ipynb](url) 
But the results were confusing.
![self_energy_imag](https://github.com/user-attachments/assets/0fafc450-71a0-4032-a212-f6dcf88bb056)
![self_energy_real](https://github.com/user-attachments/assets/cff4b823-3d53-4b2f-b114-10d19c6d4bf5)

I guess this because there are three orbitals (dxy,dxz,dyz) in svo system.  How can I plot the self-energy in the svo system?

**Comment by the-hampel** [comment]
Hi @KeyuAN95 , 

lets start with Question 1: 
I am not sure what goes wrong here. Can you show me the lines you executed to plot the spectral function? One problem I see is that the error contains `MatsubaraFreq`, but at this point of plotting the spectral function you need a self-energy on the real frequency axis. Which self-energy did you handover to the postprocessing routines here? 

Question 2: 
Yes indeed you plot automatically all self-energy elements. You can select specific orbitals like this: 
```
for orb in range(3):
    ax.oplot(Sigma_imp_iw['up_0'][orb,orb].imag, '-o', label=str(orb))
```
The flat self-energies are from the off-diagonal elements which are typically much smaller than diagonal components. 

Best,
Alex

**Reply by KeyuAN95** [reply]
Thanks for your help. It works.

**Reply by KeyuAN95** [reply]
let me make the question 1 more clearer.
I run a one-shot DMFT using solid_dmft with the flowing dmft_config.toml:
########################
```
[general]
seedname = "svo"
jobname = "out"
enforce_off_diag = true
block_threshold = 0.001
prec_mu = 0.001
h_int_type = "kanamori"
U = 4.5
J = 0.675
mu_initial_guess = 12.3847
beta = 40
n_iter_dmft = 8
dc_type = 1
dc = true
dc_dmft = true
calc_energies = false
sigma_mix = 1.0
h5_save_freq = 1
[solver]
type = "cthyb"
n_l = 35
length_cycle = 200
n_warmup_cycles = 8000
n_cycles_tot = 4e+6
imag_threshold = 1e-5
measure_G_l = true
perform_tail_fit = false
```
#####################
Then, I plotted the self energy, analytic continuation of the self energy, and correlated bands with the following script:
################################################################
```
import os
import sys
from itertools import product
import numpy as np
np.set_printoptions(precision=4,suppress=True, linewidth=140)
import scipy
import matplotlib.pyplot as plt
from h5 import HDFArchive
from triqs.gf import *
from triqs.gf.descriptors import Fourier, Omega
from triqs.plot.mpl_interface import plt,oplot
from triqs.lattice.utils import TB_from_wannier90, k_space_path
import solid_dmft.postprocessing.plot_correlated_bands as pcb
import solid_dmft.postprocessing.maxent_sigma as sigma_maxent
from triqs_dft_tools.converters import Wannier90Converter

w90_dict = {'w90_path': '../../',
            'w90_seed': 'svo',
            'n_orb': 3,
            'mu_tb': 12.384}

plot_dict = {'colorscheme_bands': 'coolwarm'}
orbital_order_to = ['dxy', 'dxz', 'dyz']
proj_on_orb = None # or 'dxy' etc
tb_dict = {'bands_path': [('R', 'G'), ('G', 'X'), ('X', 'M'), ('M', 'G')], 'n_k': 50,
            'G': [ 0.0,  0.0,  0.0], 'R': [0.5, 0.5, 0.5], 'X': [0.5,  0.0,  0.0 ], 'M': [0.5,  0.5, 0.0] }
tb_bands, alatt_k_w_path, freq_dict = pcb.get_dmft_bands(**w90_dict, **tb_dict, eta=0.01)

band_data = np.loadtxt('../../svo_bands.dat.gnu') 
with HDFArchive('./svo.h5','r') as h5:
    print(h5)
    
    obs = h5['DMFT_results/observables']
    conv_obs = h5['DMFT_results/convergence_obs']

    Sigma_imp_iw = h5['DMFT_results/last_iter/Sigma_freq_0']
    G_imp_iw = h5['DMFT_results/last_iter/Gimp_freq_0']
    G_imp_tau = h5['DMFT_results/last_iter/Gimp_time_0']
   
    mu = h5['DMFT_results']['last_iter']['chemical_potential_post']
    dc = h5['DMFT_results']['last_iter']['DC_pot']
fig, ax = plt.subplots(nrows=6, dpi=120, figsize=(7,10), sharex=True)
ax[0].plot(obs['iteration'], obs['mu'], '-o', color='C0')
ax[0].set_ylabel(r'$\mu$ (eV)')

imp_occ = np.array(obs['imp_occ'][0]['up'])+np.array(obs['imp_occ'][0]['down'])
ax[1].plot(obs['iteration'], imp_occ, '-o', color='C1')
ax[1].set_ylim(0.92,1.02)
ax[1].set_ylabel('Imp. occupation')
Aw0 = -1*(np.array(obs['imp_gb2'][0]['up'])+np.array(obs['imp_gb2'][0]['down']))
ax[2].plot(obs['iteration'], Aw0 , '-o', color='C2')
ax[2].set_ylim(0,)
ax[2].set_ylabel(r'$\bar{A}(\omega=0$)')

Z = 0.5*(np.array(obs['orb_Z'][0]['up'])+np.array(obs['orb_Z'][0]['down']))
ax[3].plot(obs['iteration'], Z, '-o', color='C3')
ax[3].set_ylim(0,1.2)
ax[3].set_ylabel(r'QP weight Z')

ax[4].semilogy(obs['iteration'][1:], conv_obs['d_G0'][0], '-o', color='C4')
ax[4].set_ylabel(r'dG$_0$')

ax[5].semilogy(obs['iteration'][1:], conv_obs['d_Gimp'][0], '-o', color='C5')
ax[5].set_ylabel(r'|G$_{imp}$-G$_{loc}$|')

ax[-1].set_xticks(range(0,len(obs['iteration'])))
ax[-1].set_xlabel('Iterations')
plt.savefig("./convergence.png", bbox_inches='tight')
plt.close()
print(Sigma_imp_iw)
list(Sigma_imp_iw.indices)
print(Sigma_imp_iw['up_0'])

fig, ax = plt.subplots(dpi=150, figsize=(6,3))
for orb in range(3):
    ax.oplot(Sigma_imp_iw['up_0'][orb,orb].imag, '-o', label=str(orb))
ax.set_ylabel(r"Im $\Sigma (i \omega)$")
ax.set_xlim(0,10)
plt.savefig('./self_energy_imag.png', bbox_inches='tight')
plt.close()
fig, ax = plt.subplots(dpi=150, figsize=(6,3))
for orb in range(3):
    ax.oplot(Sigma_imp_iw['up_0'][orb,orb].real, '-o', label=str(orb))
    #ax.oplot(Sigma_imp_iw['up_0'].real, '-o', c='C1')
ax.set_ylabel(r"Re $\Sigma (i \omega)$")
ax.set_xlim(0,10)
plt.savefig('./self_energy_real.png',  bbox_inches='tight')
plt.close()

fig, ax = plt.subplots(dpi=150, figsize=(6,4))
ax.oplot(G_imp_tau['up_0'].real, '-', c='C0', label='raw MC data')
G_imp_rebinned = G_imp_tau['up_0'].rebinning_tau(new_n_tau=300)
ax.oplot(G_imp_rebinned.real, '-', c='C1', label='rebinned')
ax.set_ylabel(r"G$ (\tau)$")
ax.legend()
plt.savefig('Gt', bbox_inches='tight')
plt.close

# use pcb maxent script to continue sigma#
Sigma_real_freq = sigma_maxent.main(external_path='svo.h5',
                                     omega_min = -10, omega_max= 10,
                                     maxent_error=0.03, iteration= None,
                                     n_points_maxent=101,
                                     continuator_type='inversion_sigmainf')[0][0]
fig, ax = plt.subplots(1, 2, sharex=True, dpi=150, figsize=(10,4))
ax[0].oplot((Sigma_real_freq['up_0'][0,0] - Sigma_real_freq['up_0'](0.0)[0,0]).real, '-', c='C0', label='Re')
ax[1].oplot(-1*Sigma_real_freq['up_0'][0,0].imag, '-', c='C0', label='Im')
ax[0].set_ylabel(r"Re $\Sigma (\omega)$ (eV)")
ax[1].set_ylabel(r"-Im $\Sigma (\omega)$ (eV)")
ax[0].set_xlim(-0.15,0.15)
ax[0].set_ylim(-0.06,0.06)
ax[1].set_xlim(-0.15,0.15)
ax[1].set_ylim( 0.0,0.06)
plt.savefig('self-energy',  bbox_inches='tight')
plt.close
  

w_mesh = {'window': [-1.2, 1.5], 'n_w': int(1001)}
sigma_dict = {'dmft_path': 'svo.h5',
              'it': 'last_iter',
              'spin': 'up', 'block' : 0,
              'w_mesh': w_mesh}
plot_dict = {'colorscheme_bands': 'Greys',
             'colorscheme_alatt': 'Spectral_r',
             'colorscheme_kslice': 'Spectral_r',
             'vmin': 0,   #
             'vmax': 5    #
            }
tb_bands, alatt_k_w_path, freq_dict = pcb.get_dmft_bands(with_sigma='calc', **w90_dict, **tb_dict, **sigma_dict)
fig, ax = plt.subplots(1,dpi=150,figsize=(7,4))
pcb.plot_bands(fig, ax, alatt_k_w_path, tb_bands, freq_dict, n_orb=w90_dict['n_orb'], tb=True, alatt=True, **plot_dict)
plt.savefig('correlated-band',  bbox_inches='tight')
plt.close()
```
##############################################################################
I got the results like these:
![convergence](https://github.com/user-attachments/assets/e92d08b5-2756-4f2a-85d7-dc0a46e313d4)
![correlated-band](https://github.com/user-attachments/assets/0c11ce8c-c3fd-4868-848d-d456c42f53a1)
![self_energy_imag](https://github.com/user-attachments/assets/d6255f94-6ad0-4958-a681-cb5e2a63efa2)
![self_energy_real](https://github.com/user-attachments/assets/6e9880a9-5f7e-4390-bb63-267d1a8b10c0)


**The spectral function is discontinuous at the Gama point. I tried the different dmft_config.toml parameters and still not resolved it. Could you help me to find the problem?** 

Question 3:
I tried to do the analytic continuation of the self-energy.  
I'm not sure if this result is correct.
![image](https://github.com/user-attachments/assets/28fc44a6-5332-4bdc-a043-8cd93c909e18)

Question 4:
I used the following script to plot the dos.
###############
```
# triqs function to load a w90 hamiltonian as tight binding object
TB = TB_from_wannier90(seed='svo' ,path='../../')
# calculate epsilon(k) on a given mesh
# since La2CuO4 is mostly 2D, we use a 2D mesh to speed things up
k_dim = 7
k_mesh = TB.get_kmesh(n_k = (k_dim, k_dim, 1))
n_k = np.prod(k_mesh.dims)
e_k = TB.fourier(k_mesh)
# create a Mesh product of k and omega mesh:
mesh_k_w = MeshProduct(k_mesh, Sigma_real_freq.mesh)
# create a triqs Gf on a combined k and omega mesh:
G_k_w = Gf(mesh = mesh_k_w, target_shape = [3,3])
# create here first a Gf called G_loc_w for the k summation
G_loc_w = Gf(mesh = Sigma_real_freq.mesh, target_shape = [3,3])
# now we loop over k-points
for k in G_k_w.mesh.components[0]:
    # use equation above to set at each k point the lattice Gf
    G_k_w[k,:] << inverse(Omega + mu - e_k[k] - (Sigma_real_freq['up_0']-dc[0]['up']) )
    # do the summation for G_loc_w  / the factor 2 is to account for both spin channels, and we divide by the weight 1/n_k
    G_loc_w << G_loc_w + 2/n_k * inverse(Omega + mu - e_k[k] - (Sigma_real_freq['up_0']-dc[0]['up']) )
fig, ax = plt.subplots(dpi=150, figsize=(6,3))
ax.oplot(-1/np.pi*G_loc_w.imag, label=r'A($\omega$) U=3.6 eV')
# ax.plot(w90_dos[:,0] - fermi, w90_dos[:,1], label='w90 DOS')
ax.set_ylabel(r'A($\omega$)')
ax.set_xlabel(r'$\omega$')
ax.set_xlim(-5,5)
ax.set_ylim(0,1.1)
ax.legend()
plt.savefig('dos',  bbox_inches='tight')
plt.close()
```
######################
However, the dos look like this 
![image](https://github.com/user-attachments/assets/5acb4e1c-f66d-4846-a4e0-0d97531bc4c9)
How can I get the right result?

Best regards,
Keyu

**Reply by the-hampel** [reply]
Hi Keyu, 

thank you for sharing your workflow. The scripts and steps look all good to me. I would think that the analytical continuation (AC) ran into problems here. The QMC results are not super converged yet. If you could restart from what you have and change `n_iter_dmft = 4` and `n_cycles_tot = 4e+7` if possible. So just change these two parameters and rerun solid_dmft. It should tell you that is resumes the calculation. 

When finished I would like to see the output from the AC call: 
```
Sigma_real_freq = sigma_maxent.main(external_path='svo.h5',
                                     omega_min = -10, omega_max= 10,
                                     maxent_error=0.03, iteration= None,
                                     n_points_maxent=101,
                                     continuator_type='inversion_dc')[0][0]
```
and please do change `continuator_type='inversion_dc'` for testing (should both work). So please share the output from this and also please plot the self-energy again like you did before but possibly w=-1 to w=1 eV so that we can see what is going on at Gamma where the spectrum is at around -0.5 eV. 

Best,
Alex

**Reply by KeyuAN95** [reply]
Hi Alex,

Thanks for your reply.
I rerun the solid_dmft with changing n_iter_dmft = 4 and n_cycles_tot = 4e+7. I  plotted the spectral function with continuator_type='inversion_dc' and continuator_type='inversion_sigmainf'. The former can give the continuous spectral function.
![image](https://github.com/user-attachments/assets/178e872a-acf7-491a-bd62-519a8da5eba1)
![image](https://github.com/user-attachments/assets/d7726bed-8869-49d9-9304-1155ea2648d6)

If I want to plot the DOS and PDOS with the DMFT results, what should I adjust the following code:
```
# triqs function to load a w90 hamiltonian as tight binding object
TB = TB_from_wannier90(seed='svo' ,path='../../')
# calculate epsilon(k) on a given mesh
k_dim = 7
k_mesh = TB.get_kmesh(n_k = (k_dim, k_dim, 1))
n_k = np.prod(k_mesh.dims)
e_k = TB.fourier(k_mesh)
# create a Mesh product of k and omega mesh:
mesh_k_w = MeshProduct(k_mesh, Sigma_real_freq.mesh)
# create a triqs Gf on a combined k and omega mesh:
G_k_w = Gf(mesh = mesh_k_w, target_shape = [3,3])
# create here first a Gf called G_loc_w for the k summation
G_loc_w = Gf(mesh = Sigma_real_freq.mesh, target_shape = [3,3])
# now we loop over k-points
for k in G_k_w.mesh.components[0]:
    # use equation above to set at each k point the lattice Gf
    G_k_w[k,:] << inverse(Omega + mu - e_k[k] - (Sigma_real_freq['up_0']-dc[0]['up']) )
    # do the summation for G_loc_w  / the factor 2 is to account for both spin channels, and we divide by the weight 1/n_k
    G_loc_w << G_loc_w + 2/n_k * inverse(Omega + mu - e_k[k] - (Sigma_real_freq['up_0']-dc[0]['up']) )
fig, ax = plt.subplots(dpi=150, figsize=(6,3))
ax.oplot(-1/np.pi*G_loc_w.imag, label=r'A($\omega$) U=3.6 eV')
# ax.plot(w90_dos[:,0] - fermi, w90_dos[:,1], label='w90 DOS')
ax.set_ylabel(r'A($\omega$)')
ax.set_xlabel(r'$\omega$')
ax.set_xlim(-5,5)
ax.set_ylim(0,1.1)
ax.legend()
plt.savefig('dos',  bbox_inches='tight')
plt.close()
```

Best regards,
Keyu

**Reply by the-hampel** [reply]
Hi Keyu, 

nice so this looks good! `inversion_dc` is a safer choice for multi-orbital systems. If you plot the self-energies of the two MaxEnt runs on top of each other it should be easily visible why the second one gives such weird spectral function. 

For the local spectral function you also always continuate Gloc_iw itself. But okay in most cases you get more features with more k-points via continuation of Sigma and doing the ksum on the real frequency mesh as you did here. Can you be a bit more specific? Does the script above not work? 

I have some more advanced scripts that utilize MPI here: https://github.com/the-hampel/pystack/blob/master/scripts/akw_TB_lattice.py and here https://github.com/the-hampel/pystack/blob/master/scripts/dos_TB_lattice_sigma.py . But if you just want to keep it simple you can also use the post-processing of solid_dmft. If you change this part: 

```
tb_dict = {'bands_path': [('R', 'G'), ('G', 'X'), ('X', 'M'), ('M', 'G')], 'n_k': 50,
            'G': [ 0.0,  0.0,  0.0], 'R': [0.5, 0.5, 0.5], 'X': [0.5,  0.0,  0.0 ], 'M': [0.5,  0.5, 0.0] }
```
to 
```
tb_dict = {'kmesh': 'regular', 'n_k': 7}
```
and then call
```
tb_bands, alatt_loc_w, freq_dict = pcb.get_dmft_bands(with_sigma='calc', **w90_dict, **tb_dict, **sigma_dict)
```
again it will calculate it on a 7x7x7 regular k-point mesh. This call is just not mpi parallelized  right now, but for smaller mesh sizes this works fine. 

Best,
Alex

**Reply by KeyuAN95** [reply]
Hi Alex,
Thanks for your help.
Next, I tried to plot the Fermi slice with the following code,
```
fermi = 12.384
# use pcb maxent script to continue sigma
Sigma_real_freq = sigma_maxent.main(external_path='svo.h5', 
                                     omega_min = -10, omega_max= 10, 
                                     maxent_error=0.03, iteration= None,
                                     n_points_maxent=101,
                                     continuator_type='inversion_sigmainf')[0][0]

w90_dict = {'w90_path': '../../',
            'w90_seed': 'svo',
            'n_orb': 3,
            'mu_tb': fermi}
# plotting options
plot_dict = {'colorscheme_bands': 'Greys',
             'colorscheme_alatt': 'Spectral_r',
             'colorscheme_kslice': 'Spectral_r'}
tb_dict = {'bands_path': [('R', 'G'), ('G', 'X'), ('X', 'M'), ('M', 'G')], 'n_k': 50,
            'G': [ 0.0,  0.0,  0.0], 'R': [0.5, 0.5, 0.5], 'X': [0.5,  0.0,  0.0 ], 'M': [0.5,  0.5, 0.0] }

tb_kslice = {key: tb_dict[key] for key in list(tb_dict.keys()) if key.isupper()}
kslice_update = {'bands_path': [('X', 'G'),('G', 'Y')], 'Y': np.array([0.0,0.5,0.0]), 'n_k': 50, 'kz': 0.0}
tb_kslice.update(kslice_update)

# we have to specify the mesh of the spectral function and pass a new keywargs object sigma_dict
w_mesh = {'window': [-13.5, 13.5], 'n_w': int(1e6)}
sigma_dict = {'dmft_path': 'svo.h5',
              'it': 'last_iter',
              'spin': 'up',
              'block' : 0 ,
              'w_mesh': w_mesh}
tb_bands, alatt_k_w_path, freq_dict = pcb.get_dmft_bands(fermi_slice = True, with_sigma='calc', **w90_dict, **tb_kslice, **sigma_dict)

fig, ax = plt.subplots(1,dpi=500,figsize=(7,5))
pcb.plot_kslice(fig, ax, alatt_k_w_path, tb_bands, freq_dict, n_orb=w90_dict['n_orb'],
                tb_dict=tb_kslice, tb=True, alatt=True, quarter=[0,1,2,3], **plot_dict)
plt.title('kz = 0.0')
plt.savefig('Fermi_slices_kz=0.0.png')
plt.close()
```
I can get the image.
![image](https://github.com/user-attachments/assets/84cc63c3-71bb-4450-bddf-b72f24ff6b83)
But there was an error when I set the 'kz = -0.5'.
```
File "/mnt/sg001/home/fz_um_aky/strongly_correlated_electrons/svo_qe/dmft_test5_keep/out2/Fermi_slices.py", line 77, in <module>
    tb_bands, alatt_k_w_path, freq_dict = pcb.get_dmft_bands(fermi_slice = True, with_sigma='calc', **w90_dict, **tb_kslice, **sigma_dict)
                                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/mnt/sg001/home/fz_um_aky/miniconda3/envs/triqs/lib/python3.12/site-packages/solid_dmft/postprocessing/plot_correlated_bands.py", line 844, in get_dmft_bands
    assert 'Z' in specs, 'Please provide Z point coordinate in tb_data_dict as input coordinate'
           ^^^^^^^^^^^^
AssertionError: Please provide Z point coordinate in tb_data_dict as input coordinate
```
How can I deal with this error?

**Reply by phibeck** [reply]
Hi Keyu,

the code needs to know 0.5 along which direction, so you need to provide the coordinates of the Z point. You just need to add it here and adjust the coordinates depending on the geometry of your lattice of course. Below I assume this is still cubic (where Z is identical with X).

```
tb_dict = {'bands_path': [('R', 'G'), ('G', 'X'), ('X', 'M'), ('M', 'G')], 'n_k': 50,
            'G': [ 0.0,  0.0,  0.0], 'R': [0.5, 0.5, 0.5], 'X': [0.5,  0.0,  0.0 ], 'M': [0.5,  0.5, 0.0] , 'Z': [0.0, 0.0, 0.5]}
```
            
Best,
Sophie

**Reply by KeyuAN95** [reply]
Thanks Sophie.
It works now.

Best,
Keyu

**Reply by KeyuAN95** [reply]
How can I adjust the parameters if I want to plot the Fermi slice when the Fermi energy is -0.5 eV?

**Comment by phibeck** [comment]
Hi, if you are only interested in the Fermi slice for the tight-binding bands, then the answer is simply that you can adjust the Fermi level in `mu_tb` in `w90_dict`. If you want to adjust the Fermi energy for the spectral function, I am afraid this is not implemented yet, but wouldn't be too hard to do. In the latter case, would you mind making an issue? We will try to add this feature as soon as possible.
